src/libopensc/card-masktech.c

```diff
@@ -0,0 +1,480 @@
+                                      size_t datalen,
+                                      u8 * out,
+                                      size_t outlen)
+{
+
+	struct masktech_private_data *private_data;
+	u8 sha256hash[32];
+	static const u8 hdr_sha256[] = {
+		0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65,
+		0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20
+	};
+	assert(card != NULL && data != NULL && out != NULL);
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "masktech_compute_signature()\n");
+	
+	/* retrieve the key reference */
+	private_data = (struct masktech_private_data *) card->drv_data;
+	if (!private_data)
+		return SC_ERROR_INTERNAL;
+
+	if (private_data->rsa_key_ref == 0x88)
+	{
+		/* for this key reference, the card supports only SHA256 hash and the hash is computed using a digest info */
+		/* check that it is a SHA256 with digest info*/
+		if ((datalen != sizeof(hdr_sha256) + 32) || (memcmp(hdr_sha256, data, sizeof(hdr_sha256)) != 0))
+		{
+			sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "It is not a SHA256 with digestinfo\n");
+			return SC_ERROR_NOT_SUPPORTED;
+		}
+		/* extract the SHA-256 hash */
+		memcpy(sha256hash, (u8 *)(data+(datalen-32)), 32);//only last 32 byte => sha256
+		/* default ISO 7816 functions */
+		return iso_ops->compute_signature(card, sha256hash, 32, out, outlen);
+	}
+	else
+	{
+		/* default ISO 7816 functions */
+		return iso_ops->compute_signature(card, data, datalen, out, outlen);
+	}
+}
+
+static int masktech_decipher(sc_card_t *card,
+                             const u8 * crgram,
+                             size_t crgram_len,
+                             u8 * out,
+                             size_t outlen)
+{
+	int r, cse;
+	sc_apdu_t apdu;
+	u8 rbuf[SC_MAX_EXT_APDU_BUFFER_SIZE];
+	u8 sbuf[SC_MAX_EXT_APDU_BUFFER_SIZE];
+	unsigned char use_ext;
+
+	assert(card != NULL && crgram != NULL && out != NULL);
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "masktech_decipher()\n");
+	
+	use_ext = crgram_len > 255; 
+	cse = use_ext ? SC_APDU_CASE_4_EXT : SC_APDU_CASE_4_SHORT;
+	sc_format_apdu(card, &apdu, cse, 0x2A, 0x80, 0x86);
+	apdu.resp = rbuf;
+	apdu.resplen = use_ext ? sizeof(rbuf) : 256;
+	apdu.le = use_ext ? 65536 : 256;
+
+	memcpy(&sbuf, crgram, crgram_len);
+	apdu.data = sbuf;
+	apdu.lc = crgram_len;
+	apdu.datalen = crgram_len;
+
+	card->caps |= use_ext ? SC_CARD_CAP_APDU_EXT : 0;
+
+	r = sc_transmit_apdu(card, &apdu);
+	SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "APDU transmit failed");
+	if (apdu.sw1 == 0x90 && apdu.sw2 == 0x00) {
+		size_t len = apdu.resplen > outlen ? outlen : apdu.resplen;
+
+		memcpy(out, apdu.resp, len);
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, len);
+	}
+	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, sc_check_sw(card, apdu.sw1, apdu.sw2));
+}
+
+/* we built a special APDU to be compatible with pinpad readers */
+/* maybe patch opensc to strip the first pin in iso build APDU if SC_PIN_CMD_IMPLICIT_CHANGE is set */
```

Comment:
The card doesn't support SC_PIN_CMD_CHANGE with the old and new pin at the same time.

I need p1 = 1 in https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/iso7816.c#L1009
pin1 is never set while pin2 can be set. But if a pinpad is used, the behavior change because in https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/iso7816.c#L1002. It asks for the previous one.

Maybe the test
`if (data->pin1.len != 0 || use_pin_pad) {`
can be changed to : 
`if (data->pin1.len != 0 || (use_pin_pad && !( data->flags & SC_PIN_CMD_IMPLICIT_CHANGE))) {`

Same problem with SC_PIN_CMD_UNBLOCK

But I was advised to not change everything for this driver ;-)